quinn/src/connection.rs

```diff
@@ -29,15 +29,57 @@ pub use crate::builders::{
 use crate::{ConnectionEvent, EndpointEvent};
 
 /// Connecting future
-pub struct ConnectingFuture(Option<ConnectionDriver>);
+pub struct Connecting(Option<ConnectionDriver>);
 
-impl ConnectingFuture {
+impl Connecting {
     pub(crate) fn new(conn: ConnectionRef) -> Self {
         Self(Some(ConnectionDriver(conn)))
     }
+
+    /// Convert into a 0-RTT or 0.5-RTT connection at the cost of weakened security
+    ///
+    /// Additionally returns a future that will yield the connection's [`IncomingStreams`] when the
+    /// connection is established.
+    ///
+    /// This is useful for reducing start-up latency by beginning transmission of application data
+    /// without waiting for the handshake's cryptographic security guarantees to be established.
+    ///
+    /// # Security
+    ///
+    /// On outgoing connections, this enables transmisison of 0-RTT data, which might be vulnerable
+    /// to replay attacks, and should therefore never invoke non-idempotent operations.
+    ///
+    /// On incoming connections, this enables transmission of 0.5-RTT data, which might be
+    /// intercepted by a man-in-the-middle. If this occurs, the handshake will not complete
+    /// successfully.
+    ///
+    /// Once [`Connection::is_handshaking`] returns false, further transmissions are unaffected by
+    /// these weaknesses.
+    ///
+    /// # Errors
+    ///
+    /// Outgoing connections are only 0-RTT-capable when a cryptographic session ticket cached from
+    /// a previous connection to the same server is available, and includes a 0-RTT key. If no such
+    /// ticket is found, the `Handshake` is returned unmodified.
+    ///
+    /// For incoming connections, a 0.5-RTT connection will always be successfully constructed.
+    pub fn into_0rtt(mut self) -> Result<(ConnectionDriver, Connection, IncomingStreams), Self> {
+        let can_0rtt = {
+            let conn = (self.0.as_mut().unwrap().0).lock().unwrap();
+            conn.inner.has_0rtt() & conn.inner.is_handshaking()
```

Comment:
```suggestion
            conn.inner.has_0rtt() && conn.inner.is_handshaking()
```

Also, do we really need to fail this if the handshake has already completed? Seems like succeeding there would be harmless.

Reply:
Hmm, I guess not.